### PR TITLE
Implement smart buffer management.

### PR DESCRIPTION
**Pull request by yuslepukhin**
Introduce a new DBOption random_access_max_buffer_size to limit
  the size of the random access buffer used for unbuffered access.
  Implement read ahead buffering when enabled.
  To that effect propagate compaction_readahead_size and the new option
  to the env options to make it available for the implementation.
  Add Hint() override so SetupForCompaction() call would call Hint()
  readahead can now be setup from both Hint() and EnableReadAhead()
  Add new option random_access_max_buffer_size support
  db_bench, options_helper to make it string parsable
  and the unit test.

### CODE REVIEW

**Comment by yuslepukhin** [comment]
@siying This is a continuation of our discussion in the past. per your request I added a comment to the new option that it is only supported on Windows. I also left the call to EnableReadAhead() at the ReadaheadRandomAccess(). I also introduced Hint() override so SetupForCompaction() would have an effect of enabling read ahead. With the new table reader option set this provides a great deal of speed improvement in our tests.

File: db/db_bench.cc [hunk 1]
```diff
@@ -373,6 +373,8 @@ DEFINE_int32(new_table_reader_for_compaction_inputs, true,
 
 DEFINE_int32(compaction_readahead_size, 0, "Compaction readahead size");
 
+DEFINE_int32(random_access_max_buffer_size, 1024 * 1024, "Maximum windows randomaccess buffer size");
```
Comment (siying) [inline db/db_bench.cc]: Does it exceed 80 characters?

**Comment by siying** [comment]
I didn't review port/win/env_win.cc. I hope you find a teammate to review it. If you fix the 80-char issue and say Windows codes are OK, I'm going to merge it.

**Comment by yuslepukhin** [comment]
@siying Excellent.

**Comment by siying** [comment]
I'm confused. You closed it. Did anyone merge anything?

**Comment by yuslepukhin** [comment]
@siying My mistake. Thought it is already merged.

**Pull request closed without merging**